src/main/java/com/pokegoapi/api/gym/Battle.java

```diff
@@ -0,0 +1,186 @@
+ *     (at your option) any later version.
+ *
+ *     This program is distributed in the hope that it will be useful,
+ *     but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *     GNU General Public License for more details.
+ *
+ *     You should have received a copy of the GNU General Public License
+ *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.pokegoapi.api.gym;
+
+import POGOProtos.Data.Battle.BattleActionOuterClass.BattleAction;
+import POGOProtos.Data.Battle.BattleActionTypeOuterClass;
+import POGOProtos.Data.Battle.BattleParticipantOuterClass.BattleParticipant;
+import POGOProtos.Data.Battle.BattlePokemonInfoOuterClass.BattlePokemonInfo;
+import POGOProtos.Data.PokemonDataOuterClass;
+import POGOProtos.Networking.Requests.Messages.AttackGymMessageOuterClass;
+import POGOProtos.Networking.Requests.Messages.StartGymBattleMessageOuterClass;
+import POGOProtos.Networking.Requests.Messages.StartGymBattleMessageOuterClass.StartGymBattleMessage.Builder;
+import POGOProtos.Networking.Requests.RequestTypeOuterClass;
+import POGOProtos.Networking.Responses.AttackGymResponseOuterClass.AttackGymResponse;
+import POGOProtos.Networking.Responses.StartGymBattleResponseOuterClass.StartGymBattleResponse;
+import POGOProtos.Networking.Responses.StartGymBattleResponseOuterClass.StartGymBattleResponse.Result;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.pokegoapi.api.PokemonGo;
+import com.pokegoapi.api.pokemon.Pokemon;
+import com.pokegoapi.exceptions.LoginFailedException;
+import com.pokegoapi.exceptions.RemoteServerException;
+import com.pokegoapi.main.ServerRequest;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Battle {
+	private Gym gym;
+	private Pokemon[] team;
+	private List<BattlePokemonInfo> bteam;
+	private StartGymBattleResponse battleResponse;
+	private PokemonGo api;
+	private int active_index;
+	private Pokemon active;
+	private int defeated;
+
+	public Battle(PokemonGo api, Pokemon[] team, Gym gym) {
+		this.team = team;
+		this.gym = gym;
+		this.api = api;
+		this.active = team[0];
+		this.defeated = 0;
+		this.active_index = 0;
+
+		this.bteam = new ArrayList<BattlePokemonInfo>();
+		for (int i = 0; i < team.length; i++) {
+			bteam.add(this.createBattlePokemon(team[i]));
+		}
+
+	}
+
+	public Result start() throws LoginFailedException, RemoteServerException {
+
+		Builder builder = StartGymBattleMessageOuterClass.StartGymBattleMessage.newBuilder();
+
+		for (int i = 0; i < team.length; i++) {
+			builder.addAttackingPokemonIds(team[i].getId());
+		}
+
+
+		List<PokemonDataOuterClass.PokemonData> defenders = gym.getDefendingPokemon();
+		builder.setGymId(gym.getId());
+		builder.setPlayerLongitude(api.getLongitude());
+		builder.setPlayerLatitude(api.getLatitude());
+		builder.setDefendingPokemonId(defenders.get(0).getId()); // may need to be sorted
```

Comment:
they are already from what i saw, but you may check if the gym is actually defended (no team has claim it), its could lead to a NPE